File: labs/manipulacao/src/go/shell.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort" // @DROP
	"strconv"
	"strings"
)

func getMen(vet []int) []int {
	// @DROP
	var unique []int
	for _, elem := range vet {
		if elem > 0 {
			unique = append(unique, elem)
		}
	}
	return unique
	// @UNC
	// _ = vet
	// return nil
}

func getCalmWomen(vet []int) []int {
	// @DROP
	var unique []int
	for _, elem := range vet {
		if elem < 0 && abs(elem) < 10 {
			unique = append(unique, elem)
		}
	}
	return unique
	// @UNC
	// _ = vet
	// return nil
}

func sortVet(vet []int) []int {
	// @DROP
	sort.Ints(vet)
	return vet
	// @UNC
	// _ = vet
	// return nil
}

func sortStress(vet []int) []int {
	// @DROP
	sort.Slice(vet, func(i, j int) bool {
		return abs(vet[i]) < abs(vet[j])
	})
	return vet
	// @UNC
	// _ = vet
	// return nil
}

func reverse(vet []int) []int {
	// @DROP
	other := make([]int, len(vet))
	copy(other, vet)
	i := 0
	j := len(other) - 1
	for i < j {
		other[i], other[j] = other[j], other[i]
		i += 1
		j -= 1
	}
	return other
	// @UNC
	// _ = vet
	// return nil
}

func unique(vet []int) []int {
	// @DROP
	var uniq []int
	for _, elem := range vet {
		if !exists(uniq, elem) {
			uniq = append(uniq, elem)
		}
	}
	return uniq
	// @UNC
	// _ = vet
	// return nil
}

func repeated(vet []int) []int {
	// @DROP
	var uniq []int
	var rep []int
	for _, elem := range vet {
		if !exists(uniq, elem) {
			uniq = append(uniq, elem)
		} else {
			rep = append(rep, elem)
		}
	}
	return rep
	// @UNC
	// _ = vet
	// return nil
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	for {
		if !scanner.Scan() {
			break
		}
		fmt.Print("$")
		line := scanner.Text()
		args := strings.Split(line, " ")
		fmt.Println(line)

		switch args[0] {
		case "get_men":
			printVec(getMen(str2vet(args[1])))
		case "get_calm_women":
			printVec(getCalmWomen(str2vet(args[1])))
		case "sort":
			printVec(sortVet(str2vet(args[1])))
		case "sort_stress":
			printVec(sortStress(str2vet(args[1])))
		case "reverse":
			array := str2vet(args[1])
			other := reverse(array)
			printVec(array)
			printVec(other)
		case "unique":
			printVec(unique(str2vet(args[1])))
		case "repeated":
			printVec(repeated(str2vet(args[1])))
		case "end":
			return
		}
	}
}

func printVec(vet []int) {
	fmt.Print("[")
	for i, val := range vet {
		if i > 0 {
			fmt.Print(", ")
		}
		fmt.Print(val)
	}
	fmt.Println("]")
}

func str2vet(s string) []int {
	if s == "[]" {
		return nil
	}
	s = s[1 : len(s)-1]
	parts := strings.Split(s, ",")
	var vet []int
	for _, part := range parts {
		n, _ := strconv.Atoi(part)
		vet = append(vet, n)
	}
	return vet
}

// @DROP
func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func exists(vet []int, value int) bool {
	for _, v := range vet {
		if v == value {
			return true
		}
	}
	return false
}

File: labs/manipulacao/src/java/Shell.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class Shell {
    static List<Integer> get_men(List<Integer> vet) {
        // @DROP
        List<Integer> unique = new ArrayList<Integer>();
        for (int elem : vet) {
            if (elem > 0) {
                unique.add(elem);
            }
        }
        return unique;
        // @UNC
        // return new ArrayList<>();
    }

    static List<Integer> get_calm_women(List<Integer> vet) {
        // @DROP
        List<Integer> unique = new ArrayList<Integer>();
        for (int elem : vet) {
            if (elem < 0 && Math.abs(elem) < 10) {
                unique.add(elem);
            }
        }
        return unique;
        // @UNC
        // return new ArrayList<>();
    }

    static List<Integer> sort(List<Integer> vet) {
        // @DROP
        Collections.sort(vet);
        return vet;
        // @UNC
        // return new ArrayList<>();
    }

    static List<Integer> sort_stress(List<Integer> vet) {
        // @DROP
        Collections.sort(vet, (a, b) -> Integer.compare(Math.abs(a), Math.abs(b)));
        return vet;
        // @UNC
        // return new ArrayList<>();
    }

    static List<Integer> reverse(List<Integer> vet) {
        // @DROP
        var other = new ArrayList<Integer>();
        for (int i = vet.size() - 1; i >= 0; i--) {
            other.add(vet.get(i));
        }
        return other;
        // @UNC
        // return new ArrayList<>();
    }

    static boolean exists(List<Integer> vet, int value) {
        // @DROP
        return vet.contains(value);
        // @UNC
        // return false;
    }

    static List<Integer> unique(List<Integer> vet) {
        // @DROP
        List<Integer> unique = new ArrayList<Integer>();
        for (int elem : vet) {
            if (!exists(unique, elem)) {
                unique.add(elem);
            }
        }
        return unique;
        // @UNC
        // return new ArrayList<>();
    }

    static List<Integer> repeated(List<Integer> vet) {
        // @DROP
        List<Integer> unique = new ArrayList<Integer>();
        List<Integer> repeated = new ArrayList<Integer>();

        for (int elem : vet) {
            if (!exists(unique, elem)) {
                unique.add(elem);
            } else {
                repeated.add(elem);
            }
        }
        return repeated;
        // @UNC
        // return new ArrayList<>();
    }

    public static void main(String[] _args) {
        while (true) {
            var line = scanner.nextLine();
            var args = line.split(" ");
            var cmd = args[0];
            System.out.println("$" + line);

            if (cmd.equals("end")) {
                break;
            } else if (cmd.equals("get_men")) {
                var array = strToVet(args[1]);
                System.out.println("" + get_men(array));
            } else if (cmd.equals("get_calm_women")) {
                var array = strToVet(args[1]);
                System.out.println("" + get_calm_women(array));
            } else if (cmd.equals("sort")) {
                var array = strToVet(args[1]);
                System.out.println("" + sort(array));
            } else if (cmd.equals("sort_stress")) {
                var array = strToVet(args[1]);
                System.out.println("" + sort_stress(array));
            } else if (cmd.equals("reverse")) {
                var array = strToVet(args[1]);
                var other = reverse(array);
                System.out.println("" + array);
                System.out.println("" + other);
            } else if (cmd.equals("exists")) {
                var array = strToVet(args[1]);
                var value = Integer.parseInt(args[2]);
                System.out.println("" + exists(array, value));
            } else if (cmd.equals("unique")) {
                var array = strToVet(args[1]);
                System.out.println("" + unique(array));
            } else if (cmd.equals("repeated")) {
                var array = strToVet(args[1]);
                System.out.println("" + repeated(array));
            } else {
                System.out.println("Invalid command");
            }
        }
    }

    public static List<Integer> strToVet(String s) {
        if (s.equals("[]"))
            return new ArrayList<>();
        List<String> parts = Arrays.asList(s.substring(1, s.length() - 1).split(","));
        List<Integer> result = new ArrayList<>();
        for (String part : parts)
            result.add(Integer.parseInt(part.trim()));
        return result;
    }

    static Scanner scanner = new Scanner(System.in);
}